File: leetcode/Spiral-Matrix-ii.py
class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        empty = [[0]*n for _ in range(n)]
        visited = [[False]*n for _ in range(n)]
        dr, dc = [0,1,0,-1], [1,0,-1,0] # 오,아,왼,위
        r = c = di = 0 # (0,0)에서 시작. 방향도 오른쪽부터
        
        for i in range(1,(n*n)+1):
            empty[r][c] = i
            visited[r][c] = True
            nr, nc = r+dr[di], c+dc[di]
            
            if 0 <= nr < n and 0 <= nc < n and not visited[nr][nc]:
                r, c= nr, nc
            else:
                di = (di + 1) % 4
                r += dr[di]
                c += dc[di]
        return empty

File: By-Swift/By-Swift/Leetcode/spiral-matrix-ii.swift
class Solution {
    func generateMatrix(_ n: Int) -> [[Int]] {
        let dx = [0,1,0,-1]
        let dy = [1,0,-1,0]
        
        var result = Array(repeating: Array(repeating: 0, count: n), count: n)
        var cx = 0
        var cy = 0
        var direction = 0
        
        for i in 1...(n*n) {
            result[cx][cy] = i
            let nx = cx + dx[direction]
            let ny = cy + dy[direction]
            
            if (0 <= nx && nx < n) && (0 <= ny && ny < n) && (result[nx][ny] == 0) {
                cx = nx
                cy = ny
            } else {
                direction = (direction+1)%4
                cx += dx[direction]
                cy += dy[direction]
            }
        }
        return result
    }
}